ts: Migrate `stream_data.js` module to TypeScript.

Lalit3716 opened the pull request:
Migrates `stream_data` module to TypeScript and also refines the existing types for stream and subscription objects present in `sub_store` module.

The main crux of this module is `create_sub_from_server_data` which generates `sub` objects which we use in web client from the data we get from the server.

The type for it's parameter `attrs` which I called `StreamAttrs` is defined as union of three types:-
1)  `NeverSubscribedStream & Pick<ExtraStreamAttrs, "subscribed" | "previously_subscribed">`
2) `ApiStreamSubscription & Pick<ExtraStreamAttrs, "subscribed" | "previously_subscribed">`

Both of these types are inferred by me by looking at the `populate_subscriptions` method which calls this function with attrs confirming to the types mentioned above.

3) `Stream & {stream_weekly_traffic: number | null; subscribers: number[]; subscribed: boolean}`

This type is inferred from the `create_streams` function which calls this function with attrs having the above type.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Lalit3716 commented:
@zulipbot add "mentor review" "buddy review"

andersk commented on web/src/stream_data.ts:
```diff
@@ -61,47 +92,47 @@ class BinaryDict {
-    get(k) {
-        const res = this.trues.get(k);
+    get(k: number): T {
+        const res = this.trues.get(k.toString());
```
Which raises the question: why is `BinaryDict` using case-folding `FoldDict`s to store numeric keys? We should replace them with `Map`s.

Lalit3716 replied:
Yeah I had the same question. Changed `FoldDict` to `Map` in a separate commit.

andersk commented on web/src/stream_data.ts:
```diff
@@ -572,50 +622,54 @@ export function can_post_messages_in_stream(stream) {
     if (
-        stream.stream_post_policy === stream_post_policy_values.non_new_members.code &&
+        stream.stream_post_policy ===
+            (stream_post_policy_values.non_new_members.code as StreamPostPolicy) &&
```
Let’s fix the underlying definition of `code` so we don’t need casts like this.

```diff
-        code: 3,
+        code: StreamPostPolicy.RESTRICT_NEW_MEMBERS
```

andersk commented on web/src/stream_data.ts:
```diff
@@ -572,50 +622,54 @@ export function can_post_messages_in_stream(stream) {
     }
 
-    if (stream.stream_post_policy === stream_post_policy_values.moderators.code) {
+    if (
+        stream.stream_post_policy ===
+        (stream_post_policy_values.moderators.code as StreamPostPolicy)
+    ) {
         return false;
     }
 
     if (
         page_params.is_guest &&
-        stream.stream_post_policy !== stream_post_policy_values.everyone.code
+        stream.stream_post_policy !== (stream_post_policy_values.everyone.code as StreamPostPolicy)
     ) {
         return false;
     }
 
     const person = people.get_by_user_id(people.my_current_user_id());
-    const current_datetime = new Date(Date.now());
-    const person_date_joined = new Date(person.date_joined);
+    const current_datetime = new Date(Date.now()).getTime();
+    const person_date_joined = new Date(person.date_joined).getTime();
     const days = (current_datetime - person_date_joined) / 1000 / 86400;
     if (
-        stream.stream_post_policy === stream_post_policy_values.non_new_members.code &&
+        stream.stream_post_policy ===
+            (stream_post_policy_values.non_new_members.code as StreamPostPolicy) &&
         days < page_params.realm_waiting_period_threshold
     ) {
         return false;
     }
     return true;
 }
 
-export function is_subscribed_by_name(stream_name) {
+export function is_subscribed_by_name(stream_name: string): boolean | undefined {
     const sub = get_sub(stream_name);
-    return sub !== undefined && sub.subscribed;
+    return sub?.subscribed;
```
It’s inconsistent to do this to `is_subscribed_by_name`, `is_subscribed`, `is_web_public` but not `is_invite_only_by_stream_name`, `is_web_public_by_stream_name`.

It seems preferable to return `boolean` here rather than force callers to deal with `boolean | undefined`.

andersk commented on web/src/peer_data.ts:
```diff
@@ -82,8 +82,8 @@ export function get_subscribers(stream_id: number): number[] {
 
-export function set_subscribers(stream_id: number, user_ids: number[]): void {
-    const subscribers = new LazySet(user_ids);
+export function set_subscribers(stream_id: number, user_ids?: number[]): void {
+    const subscribers = new LazySet(user_ids ?? []);
```
I don’t think we’re expecting callers to pass `undefined` here; if they do, we should fix the callers.

andersk commented on web/src/sub_store.ts:
```diff
@@ -1,12 +1,15 @@
 import * as blueslip from "./blueslip";
 
+type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
```
Could be simplified to `Omit<T, K> & Partial<T>`.

andersk commented on web/src/sub_store.ts:
```diff
@@ -40,6 +50,21 @@ export type StreamSubscription = (Stream & StreamSpecificNotificationSettings) &
+};
+
+// This is the actual type of subscription objects we use in the app.
+export type StreamSubscription = PartialBy<
+    ApiStreamSubscription,
+    "pin_to_top" | "email_address" | "in_home_view"
+> &
+    ExtraStreamAttrs;
```
If we’re creating subscription objects without `pin_to_top`, `email_address`, or `in_home_view`, that seems like a bug?

Lalit3716 replied:
So we are for some reason also creating `sub` objects from `NeverSubscribedStream` type and never subscribed streams does not have these fields so some `sub` objects in theory will indeed not have these properties.

Related Discussion: [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/stream_data.20typescript.20migration/near/1615755).

timabbott replied:
Hmm. If it's a stream you've never been subscribed to, we know `pin_to_top=false`, and `email_address` is I think `null` -- you don't have access to it? 

For `in_home_view`, I don't think we ever want to access it -- it might be better to figure out a way to drop it from objects when first accepting them from the server in favor of the `is_muted` field that we actually use.

Lalit3716 replied:
Yeah, we can probably delete `in_home_view` the same way we delete `subscribers` ?
https://github.com/zulip/zulip/blob/2db715030d86819de8493afb879fb5c9c5e43139/web/src/stream_data.js#L740

timabbott replied:
Yeah, that seems like a good plan.

Lalit3716 replied:
Alright dropped `in_home_view` from `ApiStreamSubscription` interface also when we will have actual zod schemas for these types, the zod will automatically remove these fields for us.

andersk commented on web/src/stream_data.ts:
```diff
@@ -744,29 +798,20 @@ export function create_sub_from_server_data(attrs) {
-        description: "",
-        rendered_description: "",
-        first_message_id: attrs.first_message_id,
+        color: "color" in attrs ? attrs.color : color_data.pick_color(),
```
`pick_color` has a side effect of modifying `color_data.unused_colors`, so we shouldn’t speculatively call it here to build the defaults.

Lalit3716 replied:
It should only be called when we indeed don't have a "color" property present in `attrs` object otherwise it will use the given value.

Also in previous version of code we were setting `sub.color` and calling `pick_color` just after creating this `sub` object, like this and I think the current ternary I wrote basically does the same thing right?  
https://github.com/zulip/zulip/blob/6632eca2dcf2051043e95348658d6da02305669f/web/src/stream_data.js#L762-L764

PIG208 commented on web/src/stream_data.ts:
```diff
@@ -720,7 +740,7 @@ export function clean_up_description(sub) {
+export function create_sub_from_server_data(attrs: StreamAttrs): StreamSubscription {
     if (!attrs.stream_id) {
         // fail fast
         throw new Error("We cannot create a sub without a stream_id");
```
Down below there is a line `delete attrs.subscribers`. If we avoid that, `subscribers` do not need to be optional.

Lalit3716 replied:
I think `subscribers` should be optional here because the type for `ApiStreamSubscription` does in fact have `subscribers` as an optional property.
https://github.com/zulip/zulip/blob/6632eca2dcf2051043e95348658d6da02305669f/zerver/lib/types.py#L198.

Also we are not setting it to a default of maybe `[]` while making `attrs` for `create_sub_from_server_data` here - 
https://github.com/zulip/zulip/blob/6632eca2dcf2051043e95348658d6da02305669f/web/src/stream_data.js#L834-L841

PIG208 replied:
Hmm, that makes sense. Since we are deleting it from `attrs` before expanding it to `sub`, do we want to remove `subscribers` from `StreamSubscription`?

Lalit3716 replied:
Yeah it makes sense to remove it from `StreamSubscription` 👍

PIG208 commented on web/src/stream_data.ts:
```diff
@@ -826,9 +846,13 @@ export function initialize(params) {
+    ): void {
         for (const sub of subs) {
-            const attrs = {
+            const attrs: StreamAttrs = {
```
I found the name `StreamAttrs` a bit confusing, and it is not necessary to define `attrs` prior to calling `create_sub_from_server_data`.

We can inline this type at `create_sub_from_server_data` so there is no need to define a separate type for this.

Lalit3716 replied:
Yeah we can inline this but I thought `StreamAttrs` was a little complex to look at first so I thought it maybe a good idea to define it outside function for better readability?

PIG208 replied:
That's reasonable too, but maybe with a better name like `StreamServerData`?

timabbott replied:
I think this being a name like `ApiGenericStreamSubscription`  which is a union of types like `ApiNeverSubscribedStreamSubscription` might be a lot more readable?

Or maybe this doesn't need a type -- shouldn't it be inferred from the `subs` parameter anyway?

Lalit3716 replied:
I like the name `ApiGenericStreamSubscription`, so changed it to this name for now.

Lalit3716 commented on web/src/sub_store.ts:
```diff
@@ -40,6 +50,21 @@ export type StreamSubscription = (Stream & StreamSpecificNotificationSettings) &
     subscribers?: number[];
 };
```
So we should use zod to parse these objects right?
https://github.com/zulip/zulip/blob/6632eca2dcf2051043e95348658d6da02305669f/web/src/stream_data.js#L811-L822

PIG208 replied:
Yeah I agree.

Lalit3716 commented:
Thanks for the reviews on this one @andersk and @PIG208 I have addressed almost every comment and left some questions.

Also, I have updated the description of the PR with some important information about some types involved in this PR just so that we all are on same page.

This is ready for another look from you guys now :)

PIG208 commented on web/src/stream_data.ts:
```diff
@@ -520,27 +549,27 @@ export function can_toggle_subscription(sub) {
 }
 
-export function can_preview(sub) {
+export function can_preview(sub: StreamSubscription): boolean | undefined {
```
We can do the same thing with `can_preview` returning `boolean` consistently.

PIG208 commented on web/src/stream_data.ts:
```diff
@@ -8,10 +8,30 @@ import * as people from "./people";
+type StreamAttrs =
+    | (NeverSubscribedStream & Pick<ExtraStreamAttrs, "subscribed" | "previously_subscribed">)
+    | (ApiStreamSubscription & Pick<ExtraStreamAttrs, "subscribed" | "previously_subscribed">)
+    | (Stream & {stream_weekly_traffic: number | null; subscribers: number[]; subscribed: boolean}); // This type is used from `create_streams` function.
```
Because `NeverSubscribedStream`, `ApiStreamSubscription`, `Stream` are all extended with `subscribed`, and that `previously_subscribed` is not optional for the first two types while defined optional on `ExtraStreamAttrs`, we can perhaps refactor the type like this:
```
    (
      | (NeverSubscribedStream & {previously_subscribed: boolean})
      | (ApiStreamSubscription & {previously_subscribed: boolean})
      | (Stream & {stream_weekly_traffic: number | null; subscribers: number[]})
    ) & {subscribed: boolean};
```

timabbott replied:
Yeah that seems a bit cleaner.

timabbott replied:
Would it make sense for `previously_subscribed` to be an always-present parameter? I guess it doesn't have a clear meaning for a stream you're currently subscribed to, so maybe that's a bad idea.

Lalit3716 replied:
That refactored type looks much better to me thanks @PIG208, as for `previously_subscribed` field I am not so sure about making it non-optional.

Lalit3716 replied:
Yeah this sounds good to me, did this refactoring in a separate commit 👍

PIG208 commented:
I did a pass of the types to check with the ones returned from the server. They seem fine to me.

timabbott commented on web/src/stream_data.js:
```diff
@@ -740,29 +740,20 @@ export function create_sub_from_server_data(attrs) {
-        name: attrs.name,
-        render_subscribers: !page_params.realm_is_zephyr_mirror_realm || attrs.invite_only === true,
-        subscribed: true,
+        render_subscribers: !page_params.realm_is_zephyr_mirror_realm || attrs.invite_only,
```
Hmm, are you sure the `=== true` can be safely removed? I would worry it was there because of a concern that we wanted a value of `true|false` here, and not some other falsey value like `undefined` or `null` which `attrs.invite_only` could have had as a value.

Lalit3716 replied:
I checked all the possible code paths up until this line and it seems `invite_only` should always exist on `attrs` type.
But to be extra safe we can leave `=== true` in there but once we have more concrete `zod` schemas we can revisit this.

timabbott commented on web/src/stream_data.js:
```diff
@@ -33,8 +33,8 @@ class BinaryDict {
-    trues = new FoldDict();
-    falses = new FoldDict();
+    trues = new Map();
+    falses = new Map();
```
This commit should note that 91d32d9f168b878acf536178a2d40131cb337bc0 is what made this possible, right?

Lalit3716 replied:
Yeah mentioned this commit in the message 👍

timabbott commented on web/src/fold_dict.ts:
```diff
@@ -48,7 +48,7 @@ export class FoldDict<V> {
     }
 
-    *values(): Iterator<V> {
+    *values(): IterableIterator<V> {
```
Why this change -- I thought we made `BinaryDict` no longer use `FoldDict`? Maybe worth putting in a separate commit if it's a bug fix.

timabbott commented:
I posted a few comments; overall looks pretty close and I'd like us to avoid getting this blocked on fiddling with my ideas like https://github.com/zulip/zulip/pull/26381#discussion_r1287816985 and https://github.com/zulip/zulip/pull/26381#discussion_r1287816369; those sorts of naming things can easily be follow-ups.

Lalit3716 commented:
@zulipbot add "integration review"

timabbott commented:
Squashed the commits tagged to be squashed and merged, thanks for all the work on this @Lalit3716!
